// rpc-supervisor/lib/system.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const supervisor_js_1 = __importDefault(require("./supervisor.js"));
const api_1 = __importDefault(require("./api"));
class SystemApi extends api_1.default {
    constructor() {
        super(...arguments);
        this.basename = 'system';
    }
    createSupervisorApi() {
        return new supervisor_js_1.default(this.client);
    }
    listMethods() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.call('listMethods');
        });
    }
    methodHelp(name) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.call('methodHelp', [name]);
        });
    }
    methodSignature(name) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.call('methodSignature', [name]);
        });
    }
}
exports.default = SystemApi;
