// rpc-supervisor/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Signal = exports.StateName = exports.StateCode = exports.ProcessInfoStates = exports.ProcessInfoStateNames = void 0;
// Enums
var ProcessInfoStateNames;
(function (ProcessInfoStateNames) {
    ProcessInfoStateNames["Stopped"] = "STOPPED";
    ProcessInfoStateNames["Starting"] = "STARTING";
    ProcessInfoStateNames["Running"] = "RUNNING";
    ProcessInfoStateNames["Backoff"] = "BACKOFF";
    ProcessInfoStateNames["Stopping"] = "STOPPING";
    ProcessInfoStateNames["Exited"] = "EXITED";
    ProcessInfoStateNames["Fatal"] = "FATAL";
    ProcessInfoStateNames["Unknown"] = "UNKNOWN";
})(ProcessInfoStateNames = exports.ProcessInfoStateNames || (exports.ProcessInfoStateNames = {}));
var ProcessInfoStates;
(function (ProcessInfoStates) {
    ProcessInfoStates[ProcessInfoStates["Stopped"] = 0] = "Stopped";
    ProcessInfoStates[ProcessInfoStates["Starting"] = 10] = "Starting";
    ProcessInfoStates[ProcessInfoStates["Running"] = 20] = "Running";
    ProcessInfoStates[ProcessInfoStates["Backoff"] = 30] = "Backoff";
    ProcessInfoStates[ProcessInfoStates["Stopping"] = 40] = "Stopping";
    ProcessInfoStates[ProcessInfoStates["Exited"] = 100] = "Exited";
    ProcessInfoStates[ProcessInfoStates["Fatal"] = 200] = "Fatal";
    ProcessInfoStates[ProcessInfoStates["Unknown"] = 1000] = "Unknown";
})(ProcessInfoStates = exports.ProcessInfoStates || (exports.ProcessInfoStates = {}));
var StateCode;
(function (StateCode) {
    StateCode[StateCode["Fatal"] = 2] = "Fatal";
    StateCode[StateCode["Running"] = 1] = "Running";
    StateCode[StateCode["Restarting"] = 0] = "Restarting";
    StateCode[StateCode["Shutdown"] = -1] = "Shutdown";
})(StateCode = exports.StateCode || (exports.StateCode = {}));
var StateName;
(function (StateName) {
    StateName["Fatal"] = "FATAL";
    StateName["Running"] = "RUNNING";
    StateName["Restarting"] = "RESTARTING";
    StateName["Shutdown"] = "SHUTDOWN";
})(StateName = exports.StateName || (exports.StateName = {}));
var Signal;
(function (Signal) {
    Signal["One"] = "1";
    Signal["Hup"] = "HUP";
})(Signal = exports.Signal || (exports.Signal = {}));
