// ngx-editor/esm2020/lib/editor.service.mjs
import { Injectable, Optional } from '@angular/core';
import Locals from './Locals';
import * as i0 from "@angular/core";
import * as i1 from "./editor-config.service";
export class NgxEditorService {
    constructor(config) {
        this.config = config;
    }
    get locals() {
        return new Locals(this.config.locals);
    }
}
NgxEditorService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.5", ngImport: i0, type: NgxEditorService, deps: [{ token: i1.NgxEditorServiceConfig, optional: true }], target: i0.ɵɵFactoryTarget.Injectable });
NgxEditorService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.2.5", ngImport: i0, type: NgxEditorService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.5", ngImport: i0, type: NgxEditorService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return [{ type: i1.NgxEditorServiceConfig, decorators: [{
                    type: Optional
                }] }]; } });
export const provideMyServiceOptions = (config) => {
    return {
        locals: config.locals ?? {},
    };
};
