// ng-select2-component/esm2015/lib/select2-interfaces.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/select2-interfaces.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function Select2Group() { }
if (false) {
    /**
     * label of group
     * @type {?}
     */
    Select2Group.prototype.label;
    /**
     * options list
     * @type {?}
     */
    Select2Group.prototype.options;
    /**
     * add classes
     * @type {?|undefined}
     */
    Select2Group.prototype.classes;
    /**
     * template id
     * @type {?|undefined}
     */
    Select2Group.prototype.templateId;
    /**
     * template data
     * @type {?|undefined}
     */
    Select2Group.prototype.data;
}
/**
 * @record
 */
export function Select2Option() { }
if (false) {
    /**
     * value
     * @type {?}
     */
    Select2Option.prototype.value;
    /**
     * label of option
     * @type {?}
     */
    Select2Option.prototype.label;
    /**
     * no selectable is disabled
     * @type {?|undefined}
     */
    Select2Option.prototype.disabled;
    /**
     * for identification
     * @type {?|undefined}
     */
    Select2Option.prototype.id;
    /**
     * add classes
     * @type {?|undefined}
     */
    Select2Option.prototype.classes;
    /**
     * template id
     * @type {?|undefined}
     */
    Select2Option.prototype.templateId;
    /**
     * template data
     * @type {?|undefined}
     */
    Select2Option.prototype.data;
    /**
     * hide this option
     * @type {?|undefined}
     */
    Select2Option.prototype.hide;
}
/**
 * @record
 * @template U
 */
export function Select2UpdateEvent() { }
if (false) {
    /** @type {?} */
    Select2UpdateEvent.prototype.component;
    /** @type {?} */
    Select2UpdateEvent.prototype.value;
    /** @type {?} */
    Select2UpdateEvent.prototype.options;
}
/**
 * @record
 * @template U
 */
export function Select2SearchEvent() { }
if (false) {
    /** @type {?} */
    Select2SearchEvent.prototype.component;
    /** @type {?} */
    Select2SearchEvent.prototype.value;
    /** @type {?} */
    Select2SearchEvent.prototype.search;
}
/**
 * @record
 * @template U
 */
export function Select2RemoveEvent() { }
if (false) {
    /** @type {?} */
    Select2RemoveEvent.prototype.component;
    /** @type {?} */
    Select2RemoveEvent.prototype.value;
    /** @type {?} */
    Select2RemoveEvent.prototype.removedOption;
}
/**
 * @record
 */
export function Select2ScrollEvent() { }
if (false) {
    /** @type {?} */
    Select2ScrollEvent.prototype.component;
    /** @type {?} */
    Select2ScrollEvent.prototype.way;
    /** @type {?} */
    Select2ScrollEvent.prototype.search;
}
