// simplebar-angular/bundles/simplebar-angular.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('simplebar-core')) :
    typeof define === 'function' && define.amd ? define('simplebar-angular', ['exports', '@angular/core', 'simplebar-core'], factory) :
    (global = global || self, factory(global['simplebar-angular'] = {}, global.ng.core, global.SimpleBar));
}(this, (function (exports, core, SimpleBar) { 'use strict';

    SimpleBar = SimpleBar && Object.prototype.hasOwnProperty.call(SimpleBar, 'default') ? SimpleBar['default'] : SimpleBar;

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/simplebar-angular.component.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var SimplebarAngularComponent = /** @class */ (function () {
        /**
         * @param {?} elRef
         */
        function SimplebarAngularComponent(elRef) {
            this.elRef = elRef;
        }
        /**
         * @return {?}
         */
        SimplebarAngularComponent.prototype.ngOnInit = function () { };
        /**
         * @return {?}
         */
        SimplebarAngularComponent.prototype.ngAfterViewInit = function () {
            this.SimpleBar = new SimpleBar(this.elRef.nativeElement, this.options || {});
        };
        /**
         * @return {?}
         */
        SimplebarAngularComponent.prototype.ngOnDestroy = function () {
            this.SimpleBar.unMount();
            this.SimpleBar = null;
        };
        return SimplebarAngularComponent;
    }());
    SimplebarAngularComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'ngx-simplebar',
                    host: { 'data-simplebar': 'init' },
                    template: "<div class=\"simplebar-wrapper\">\n  <div class=\"simplebar-height-auto-observer-wrapper\">\n    <div class=\"simplebar-height-auto-observer\"></div>\n  </div>\n  <div class=\"simplebar-mask\">\n    <div class=\"simplebar-offset\">\n      <div class=\"simplebar-content-wrapper\">\n        <div class=\"simplebar-content\">\n          <ng-content></ng-content>\n        </div>\n      </div>\n    </div>\n  </div>\n  <div class=\"simplebar-placeholder\"></div>\n</div>\n<div class=\"simplebar-track simplebar-horizontal\">\n  <div class=\"simplebar-scrollbar\"></div>\n</div>\n<div class=\"simplebar-track simplebar-vertical\">\n  <div class=\"simplebar-scrollbar\"></div>\n</div>\n",
                    encapsulation: core.ViewEncapsulation.None,
                    styles: ["[data-simplebar]{align-content:flex-start;align-items:flex-start;flex-direction:column;flex-wrap:wrap;justify-content:flex-start;position:relative}.simplebar-wrapper{height:inherit;max-height:inherit;max-width:inherit;overflow:hidden;width:inherit}.simplebar-mask{direction:inherit;height:auto!important;overflow:hidden;width:auto!important;z-index:0}.simplebar-mask,.simplebar-offset{bottom:0;left:0;margin:0;padding:0;position:absolute;right:0;top:0}.simplebar-offset{-webkit-overflow-scrolling:touch;box-sizing:inherit!important;direction:inherit!important;resize:none!important}.simplebar-content-wrapper{-ms-overflow-style:none;box-sizing:border-box!important;direction:inherit;display:block;height:100%;max-height:100%;max-width:100%;position:relative;scrollbar-width:none;width:auto}.simplebar-content-wrapper::-webkit-scrollbar,.simplebar-hide-scrollbar::-webkit-scrollbar{height:0;width:0}.simplebar-content:after,.simplebar-content:before{content:\" \";display:table}.simplebar-placeholder{max-height:100%;max-width:100%;pointer-events:none;width:100%}.simplebar-height-auto-observer-wrapper{box-sizing:inherit!important;flex-basis:0;flex-grow:inherit;flex-shrink:0;float:left;height:100%;margin:0;max-height:1px;max-width:1px;overflow:hidden;padding:0;pointer-events:none;position:relative;width:100%;z-index:-1}.simplebar-height-auto-observer{box-sizing:inherit;display:block;height:1000%;left:0;min-height:1px;min-width:1px;opacity:0;top:0;width:1000%;z-index:-1}.simplebar-height-auto-observer,.simplebar-track{overflow:hidden;pointer-events:none;position:absolute}.simplebar-track{bottom:0;right:0;z-index:1}[data-simplebar].simplebar-dragging .simplebar-content{-ms-user-select:none;-webkit-user-select:none;pointer-events:none;user-select:none}[data-simplebar].simplebar-dragging .simplebar-track{pointer-events:all}.simplebar-scrollbar{left:0;min-height:10px;position:absolute;right:0}.simplebar-scrollbar:before{background:#000;border-radius:7px;content:\"\";opacity:0;position:absolute;transition:opacity .2s linear .5s}.simplebar-hover .simplebar-scrollbar:before,.simplebar-mouse-entered .simplebar-scrollbar:before,.simplebar-scrollbar.simplebar-visible:before,.simplebar-scrolling .simplebar-scrollbar:before{opacity:.5;transition-delay:0s;transition-duration:0s}.simplebar-track.simplebar-vertical{top:0;width:11px}.simplebar-scrollbar:before{bottom:2px;left:2px;right:2px;top:2px}.simplebar-track.simplebar-horizontal{height:11px;left:0}.simplebar-track.simplebar-horizontal .simplebar-scrollbar{bottom:0;left:0;min-height:0;min-width:10px;right:auto;top:0;width:auto}[data-simplebar-direction=rtl] .simplebar-track.simplebar-vertical{left:0;right:auto}.simplebar-dummy-scrollbar-size{-ms-overflow-style:scrollbar!important;direction:rtl;height:500px;opacity:0;overflow-x:scroll;overflow-y:hidden;position:fixed;visibility:hidden;width:500px}.simplebar-dummy-scrollbar-size>div{height:200%;margin:10px 0;width:200%}.simplebar-hide-scrollbar{-ms-overflow-style:none;left:0;overflow-y:scroll;position:fixed;scrollbar-width:none;visibility:hidden}", "ngx-simplebar{display:block}"]
                }] }
    ];
    /** @nocollapse */
    SimplebarAngularComponent.ctorParameters = function () { return [
        { type: core.ElementRef }
    ]; };
    SimplebarAngularComponent.propDecorators = {
        options: [{ type: core.Input, args: ['options',] }]
    };
    if (false) {
        /** @type {?} */
        SimplebarAngularComponent.prototype.options;
        /** @type {?} */
        SimplebarAngularComponent.prototype.elRef;
        /** @type {?} */
        SimplebarAngularComponent.prototype.SimpleBar;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/simplebar-angular.module.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var SimplebarAngularModule = /** @class */ (function () {
        function SimplebarAngularModule() {
        }
        return SimplebarAngularModule;
    }());
    SimplebarAngularModule.decorators = [
        { type: core.NgModule, args: [{
                    declarations: [SimplebarAngularComponent],
                    imports: [],
                    exports: [SimplebarAngularComponent]
                },] }
    ];

    /**
     * @fileoverview added by tsickle
     * Generated from: public-api.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * Generated from: simplebar-angular.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.SimplebarAngularComponent = SimplebarAngularComponent;
    exports.SimplebarAngularModule = SimplebarAngularModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
